examples/colors_rgb.rs

```diff
@@ -1,72 +1,130 @@
+        let mut terminal = Terminal::new(CrosstermBackend::new(writer))?;
+        let size = terminal.size()?;
+        let mut app = Self::new()?;
+        app.setup_colors(size.width, size.height * 2);
+        terminal.clear()?;
+        while !app.should_quit {
+            terminal.draw(|frame| {
+                app.render(frame);
+            })?;
+            app.handle_events()?;
+            app.frame_count += 1;
         }
         restore_terminal()?;
         Ok(())
     }
 
-    fn draw(&mut self) -> Result<()> {
-        self.terminal.draw(|frame| {
-            frame.render_widget(RgbColors, frame.size());
-        })?;
-        Ok(())
+    fn render(&mut self, frame: &mut Frame) {
+        let layout = Layout::default()
+            .direction(Direction::Vertical)
+            .constraints([Constraint::Length(1), Constraint::Min(0)])
+            .split(frame.size());
+        self.calculate_fps();
+        let title = format!(
+            "colors_rgb example. Press q to quit. {fps:.2} fps",
+            fps = self.fps
+        );
+        let title = Paragraph::new(title).alignment(Alignment::Center);
+        let rgb_colors = RgbColors {
+            colors: &self.colors,
+            frame_count: self.frame_count,
+        };
+        frame.render_widget(title, layout[0]);
+        frame.render_widget(rgb_colors, layout[1]);
+    }
+
+    fn calculate_fps(&mut self) {
+        let elapsed_frames = self.frame_count - self.last_fps_frame_count;
+        let elapsed_duration = self.last_fps_instant.elapsed();
+        // avoid noise in the FPS calculation by only updating it if at least 3 frames have passed
+        if elapsed_frames > 2 && elapsed_duration > Duration::from_secs(1) {
+            self.fps = elapsed_frames as f64 / elapsed_duration.as_secs_f64();
+            self.last_fps_frame_count = self.frame_count;
+            self.last_fps_instant = Instant::now();
+        }
     }
 
     fn handle_events(&mut self) -> Result<()> {
-        if event::poll(Duration::from_millis(100))? {
-            if let Event::Key(key) = event::read()? {
-                if key.kind == KeyEventKind::Press && key.code == KeyCode::Char('q') {
-                    self.should_quit = true;
-                };
+        if event::poll(Duration::from_millis(1))? {
+            match event::read()? {
+                Event::Key(key) => {
+                    if key.kind == KeyEventKind::Press && key.code == KeyCode::Char('q') {
+                        self.should_quit = true;
+                    };
+                }
+                Event::Resize(width, height) => {
+                    self.setup_colors(width, height * 2);
+                }
+                _ => (),
             }
         }
         Ok(())
     }
+
+    fn setup_colors(&mut self, width: u16, height: u16) {
+        self.colors.clear();
+        for y in 0..height {
+            let mut row = Vec::new();
+            for x in 0..width {
```

Comment:
I couldn't find a real fix for the crash we talked about earlier, however a workaround is to add 1 to the width. This provides to kind of security gap.

```suggestion
            for x in 0..width+1 {
```

Weirdly enough, the crash only occurs while extending the width, not the height.

I think it has to do with the window being resized between colors calculations and rendering.
If that were the case, moving `handle_event` call inside `draw` or reducing the input latency should have fixed it, but it did not so ...

Reply:
I will test the crash again with this change before merge.

Reply:
Changed the for loops to 0..=width and 0..=height